src/app/games/checkers/common/AbstractCheckersRules.ts

```diff
@@ -0,0 +1,517 @@
+import { RulesFailure } from 'src/app/jscaip/RulesFailure';
+import { ArrayUtils, MGPFallible, MGPOptional, MGPValidation } from '@everyboard/lib';
+import { CheckersMove } from './CheckersMove';
+import { CheckersFailure } from './CheckersFailure';
+import { CheckersPiece, CheckersStack, CheckersState } from './CheckersState';
+import { GameStatus } from 'src/app/jscaip/GameStatus';
+import { TableUtils } from 'src/app/jscaip/TableUtils';
+import { CoordSet } from 'src/app/jscaip/CoordSet';
+import { Localized } from 'src/app/utils/LocaleUtils';
+
+export type CheckersConfig = {
+    playerRows: number;
+    emptyRows: number;
+    width: number;
+    stackPiece: boolean;
+    maximalCapture: boolean;
+    simplePieceCanCaptureBackwards: boolean;
+    promotedPiecesCanFly: boolean;
+    occupyEvenSquare: boolean;
+}
+
+export class CheckersLocalizable {
+
+    public static readonly STACK_PIECE: Localized = () => $localize`Stack piece instead of capturing it`;
+
+    public static readonly MAXIMAL_CAPTURE: Localized = () => $localize`You must capture the highest number of piece`;
+
+    public static readonly SIMPLE_PIECE_CAN_CAPTURE_BACKWARDS: Localized = () => $localize`Simple piece can back-capture`;
+
+    public static readonly PROMOTED_PIECES_CAN_TRAVEL_LONG_DISTANCES: Localized = () => $localize`Promoted pieces can travel long distance`;
+
+    public static readonly OCCUPY_EVEN_SQUARE: Localized = () => $localize`Occupy even square`;
+
+}
+
+export class CheckersNode extends GameNode<CheckersMove, CheckersState> {}
+
+export abstract class AbstractCheckersRules extends ConfigurableRules<CheckersMove, CheckersState, CheckersConfig> {
+
+    public override getInitialState(optionalConfig: MGPOptional<CheckersConfig>): CheckersState {
+        const U: CheckersStack = new CheckersStack([CheckersPiece.ZERO]);
+        const V: CheckersStack = new CheckersStack([CheckersPiece.ONE]);
+        const _: CheckersStack = CheckersStack.EMPTY;
+        const config: CheckersConfig = optionalConfig.get();
+        const height: number = config.emptyRows + (2 * config.playerRows);
+        const board: CheckersStack[][] = TableUtils.create(config.width, height, _);
+        const occupiedSquare: number = config.occupyEvenSquare ? 0 : 1;
+        for (let y: number = 0; y < config.playerRows; y++) {
+            for (let x: number = 0; x < config.width; x++) {
+                if ((x + y) % 2 === occupiedSquare) {
+                    board[y][x] = V;
+                    board[height - 1 - y][config.width - 1 - x] = U;
+                }
+            }
+        }
+        return new CheckersState(board, 0);
+    }
+
+    /**
+     * @param state the state from which you want the current player's capture
+     * @param config the config
+     * @returns all the complete capture, wether or not they are legal
+     */
+    public getCompleteCaptures(state: CheckersState, config: CheckersConfig): CheckersMove[] {
```

Comment:
should not be public
on peut instaurer que par défaut toute méthode est privée, si c'est pas OK, elle doit alors être protected, et si c'est pas OK, alors publique ?

Reply:
Oui ! privé si présent seulement localement